File: app/src/main/java/fr/insapp/insapp/activities/SettingsActivity.java
package fr.insapp.insapp.activities;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceScreen;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import fr.insapp.insapp.R;
import fr.insapp.insapp.fragments.BarcodeSettingsFragment;
import fr.insapp.insapp.fragments.SettingsFragment;

/**
 * Created by thomas on 15/12/2016.
 *
 * Based on https://stackoverflow.com/a/36051385
 */

public class SettingsActivity extends AppCompatActivity implements PreferenceFragmentCompat.OnPreferenceStartScreenCallback {

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        // toolbar

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_settings);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                final Drawable upArrow = ContextCompat.getDrawable(SettingsActivity.this, R.drawable.abc_ic_ab_back_material);
                upArrow.setColorFilter(0xffffffff, PorterDuff.Mode.SRC_ATOP);
                getSupportActionBar().setHomeAsUpIndicator(upArrow);
            }
        }

        // handling nested preferences screen

        if (savedInstanceState == null) {
            FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
            SettingsFragment fragment = SettingsFragment.newInstance("General settings");

            fragmentTransaction.add(R.id.settings_fragment, fragment);
            fragmentTransaction.commit();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onPreferenceStartScreen(PreferenceFragmentCompat caller, PreferenceScreen preferenceScreen) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        BarcodeSettingsFragment fragment = BarcodeSettingsFragment.newInstance("Barcode settings");

        Bundle args = new Bundle();
        args.putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, preferenceScreen.getKey());
        fragment.setArguments(args);

        transaction.replace(R.id.settings_fragment, fragment, preferenceScreen.getKey());
        transaction.addToBackStack(null);
        transaction.commit();

        return true;
    }
}

File: app/src/main/java/fr/insapp/insapp/activities/SettingsActivity.kt
package fr.insapp.insapp.activities

import android.graphics.PorterDuff
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import fr.insapp.insapp.R
import fr.insapp.insapp.fragments.SettingsFragment
import kotlinx.android.synthetic.main.activity_settings.*

/**
 * Created by thomas on 15/12/2016.
 *
 * https://developer.android.com/guide/topics/ui/settings
 */

class SettingsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        // toolbar
        setSupportActionBar(toolbar_settings)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val upArrow = ResourcesCompat.getDrawable(resources, R.drawable.abc_ic_ab_back_material, null)
        upArrow?.setColorFilter(-0x1, PorterDuff.Mode.SRC_ATOP)
        supportActionBar?.setHomeAsUpIndicator(upArrow)

        supportFragmentManager
            .beginTransaction()
            .replace(R.id.settings_fragment, SettingsFragment())
            .commit()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }

            else -> {
            }
        }

        return super.onOptionsItemSelected(item)
    }
}